[DOCS] Rewrite 'rewrite' parameter docs

jrodewig opened the pull request:
### Changes
- Retitles page and removes references to multi term query. If users understand which queries are rewritten, I'm not sure if an understanding of the 'multi term query' concept is needed.
- Adds a warning noting the parameter is for expert users.
- Adds details about how impacted queries are rewritten.
- Rewrites value definitions
- Adds section for performance considerations.

This is part of #40977, an effort to standardize documentation for query types.

Any and all feedback welcome!

## Before
<details>
 <summary>Before image</summary>
<img width="777" alt="rewrite-before" src="https://user-images.githubusercontent.com/40268737/57460527-46efb280-7243-11e9-947f-66909b8a0ff7.png">
</details>

## After
<details>
 <summary>After image</summary>
<img width="777" alt="rewrite-before" src="https://user-images.githubusercontent.com/40268737/57460752-b796cf00-7243-11e9-99ed-07b99e9caffd.png">
</details>

jrodewig committed:
[DOCS] Rewrite 'rewrite' parameter docs

elasticmachine commented:
Pinging @elastic/es-search

mayya-sharipova commented on docs/reference/query-dsl/multi-term-rewrite.asciidoc:
```diff
@@ -1,45 +1,79 @@
+WARNING: This parameter is for expert users only. Changing the value of this
+parameter can impact search performance and relevance.
+
+To improve performance, {es} rewrites the following queries:
```
I am not sure if `rewrite `happens to improve performance. I think it happens just to break up complex queries into term based queries that Lucene can execute

mayya-sharipova commented on docs/reference/query-dsl/multi-term-rewrite.asciidoc:
```diff
@@ -1,45 +1,79 @@
+
+* <<query-dsl-prefix-query, `prefix`>>
+* <<query-dsl-query-string-query, `query_string`>>
+* <<query-dsl-wildcard-query, `wildcard`>>
```
There are other queries as well that get rewritten. For example, `regexp`, `fuzzy`, `prefix` can also be rewritten.

jrodewig replied:
Thanks @mayya-sharipova. I added these queries with 50505e6.

mayya-sharipova commented on docs/reference/query-dsl/multi-term-rewrite.asciidoc:
```diff
@@ -1,45 +1,79 @@
+In their original form, these queries are expensive. {es} changes the
+original query to a less expensive <<query-dsl-bool-query, `bool` query>>. This
+`bool` query contains a `should` clause and <<query-dsl-term-query, `term`
+query>> for each matching term.
```
I am not sure this is always the case, but may be @jimczi  can confirm here. As I understood it sometimes for `constant_score` instead of a boolean query,  we just use a document bit set.  And it looks liek the rewrite is done by Lucene, not ES

mayya-sharipova commented on docs/reference/query-dsl/multi-term-rewrite.asciidoc:
```diff
@@ -1,45 +1,79 @@
+=== Valid values
+
+`constant_score` (Default)::
+Automatically uses the `constant_score_boolean` or `scoring_boolean` method,
```
I don't think `constant_score` uses `scoring_boolean`

mayya-sharipova reviewed (changes requested):
Thanks @jrodewig. Very nice work, we just to need to clarify several items.

jrodewig committed:
[DOCS] Revise 'rewrite' parameter changes

jrodewig commented:
Thanks so much for your thorough feedback, @mayya-sharipova! I made several changes you suggested with 50505e6.

- Adds information about Lucene. Corrected several 'Elasticsearch' mentions to 'Lucene.'
- Adds `fuzzy` and `regexp` to the list of queries
- Added information about `constant_score` using a bit set.
- Moved information about rewriting as `bool` queries to the valid definitions

jrodewig commented on docs/reference/query-dsl/multi-term-rewrite.asciidoc:
```diff
@@ -1,45 +1,109 @@
+`constant_score` (Default)::
+Uses the `constant_score_boolean` method for fewer matching terms. Otherwise,
+this method finds all matching terms in sequence and returns matching documents
+using a bit set.
```
@mayya-sharipova  @jimczi 

Please confirm that this `constant_score` definition is accurate.

mayya-sharipova reviewed (approved):
Thanks @jrodewig. The final version looks good to me!
